net, vmispec, hotplug, tests: Use libvmi for resource creation

ronger4 opened the pull request:
Signed-off-by: ronger4 <rgershbu@redhat.com>

### What this PR does
Before this PR:
Unit tests directly created and modified VM and VMI resources by manipulating their fields.
After this PR:
Unit tests have been refactored to use the libvmi package for creating and modifying VM and VMI resources, promoting readability and standardization.
Unused funcs and vars are removed.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Partially address: https://github.com/kubevirt/kubevirt/issues/12059

### Why we need it and why it was done in this way
Using the libvmi package simplifies the unit tests and makes them more readable.
This approach ensures a standard way of creating VM and VMI resources across tests.
In addition, I removed unused vars and funcs.

The following tradeoffs were made:

- Direct manipulation of resources was replaced with utility functions, which may introduce minor overhead but enhances readability and maintainability.
- Had to add another DescribeTable as I couldn't find the implementation in libvmi for vmi.Status.Interfaces.

The following alternatives were considered:
Creating new func per Entry and creating all the resources from scratch per Entry, might be more readable but less effective and would create code dupclications.
Links to places where the discussion took place: https://github.com/kubevirt/kubevirt/issues/12059

### Special notes for your reviewer
I created this as draft for now to re-check that I didn't miss an option to implement 
`vmi.Status.Interfaces = []v1.VirtualMachineInstanceNetworkInterface{
		{Name: networkName, InterfaceName: guestIfaceName, InfoSource: vmispec.InfoSourceMultusStatus}`
Using libvmi. I didn't find such method in the libvmi package. Hence, I had to split the test to 2.  If there is an option that I missed I can align the last 2 entries as-well.

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

ronger4 commented:
Hi,

Please note my comment on
**Special notes for your reviewer** section

fossedihelm commented:
I think you can add the status interface option to libvmi/status file
```go
// WithInterfaceStatus adds a volume status
func WithInterfaceStatus(interfaceStatus v1.VirtualMachineInstanceNetworkInterface) Option {
	return func(vmiStatus *v1.VirtualMachineInstanceStatus) {
		vmiStatus.Interfaces = append(vmiStatus.Interfaces, interfaceStatus)
	}
}
```
With that you can add the status option to a vmi:
```
libvmistatus.WithStatus(
		libvmistatus.New(
			libvmistatus.WithInterfaceStatus(v1.VirtualMachineInstanceNetworkInterface{
				Name: networkName, InterfaceName: guestIfaceName, InfoSource: vmispec.InfoSourceMultusStatus,
			}),
		))
```
@EdDev for awareness

ronger4 commented:
Thanks, 
I will try adding and testing locally.

ronger4 commented:
@fossedihelm I uploaded a new commit with the code you suggested, now all tests are aligned.

fossedihelm commented:
Two initial notes:
1. Please squash the commits into a single one
2. Fill the release note in the PR description with a `NONE`

Thanks.

fossedihelm reviewed (commented):
Thank you!
Some minor comments below.

ronger4 commented:
Thanks for the extra feedback,
I have to resolve a local git issue, then I will add the extra fixes and squash the commits to a single one.

ronger4 commented:
I took care of the last notes and squashed the commits.

fossedihelm commented:
@ronger4 
I'd expect at least a change to the `pkg/network/vmispec/BUILD.bazel` file.
Can you please run `make && make generate` and amend the changes?
Thank you :)

fossedihelm commented:
/test pull-kubevirt-build

fossedihelm commented:
/test all

ronger4 commented:
I see one of the tests failed, any suggestion why?

fossedihelm commented on pkg/network/vmispec/hotplug_test.go:
```diff
@@ -22,33 +22,39 @@ package vmispec_test
-
 var _ = Describe("utilitary funcs to identify attachments to hotplug", func() {
 	const (
+		vmName         = "pepe"
```
Please drop this :)

fossedihelm reviewed (commented):
Just a leftover, other than that, it looks good from my point of view!
I'll let the sig/network have the final words.
@orelmisan @EdDev

fossedihelm commented:
> I see one of the tests failed, any suggestion why?

The failures are unrelated, and the lane is not required. All good!

ronger4 commented:
> > I see one of the tests failed, any suggestion why?
> 
> The failures are unrelated, and the lane is not required. All good!

Should I move from preview to ready for review at this point or its ok ?

fossedihelm commented:
> Should I move from preview to ready for review at this point or its ok ?

Yep! Thanks

EdDev commented on pkg/network/vmispec/hotplug_test.go:
```diff
@@ -22,33 +22,38 @@ package vmispec_test
 import (
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
-
```
This newline should be kept.

EdDev commented on pkg/network/vmispec/hotplug_test.go:
```diff
@@ -22,33 +22,38 @@ package vmispec_test
-
-	DescribeTable("NetworksToHotplugWhosePodIfacesAreReady", func(vmi *v1.VirtualMachineInstance, networksToHotplug ...v1.Network) {
+	DescribeTable("NetworksToHotplugWhosePodIfacesAreReady", func(vmiOpts []libvmi.Option, networksToHotplug ...v1.Network) {
+		vmi := libvmi.New(vmiOpts...)
```
I would prefer you pass in the VMI because it looks nicer (when we read it) than passing libvmi options.
A few more characters will not make a lot of difference. If I think about it, it will be even shorter as you replace `[]libvmi.Option{}` with `libvmi.New()`.

E.g.:
```go
Entry("VMI without networks in spec does not have anything to hotplug", libvmi.New()),
```

EdDev reviewed (commented):
Thank you, very nice.

I had one comment about what to pass to the table test and one minor nit.

EdDev reviewed (commented):
It will also be nice if you can format the comment subject a bit:

> Refactor unit tests to use libvmi for resource creation in hotplug test 

`net, vmispec, hotplug, tests: Use libvmi for resource creation`

It is just easier to track later what part of the code has changed this way, even without looking into the commit changes.

Also, if you can, please add the reason for the change as described in the PR description, in the commit text. (without that ugly "before" and "after" format)

ronger4 commented:
> It will also be nice if you can format the comment subject a bit:
> 
> > Refactor unit tests to use libvmi for resource creation in hotplug test
> 
> `net, vmispec, hotplug, tests: Use libvmi for resource creation`
> 
> It is just easier to track later what part of the code has changed this way, even without looking into the commit changes.
> 
> Also, if you can, please add the reason for the change as described in the PR description, in the commit text. (without that ugly "before" and "after" format)

I have fixed and refactored the code and commit message as you suggested.
Hopefully I didn't miss anything.

EdDev reviewed (commented):
Thank you!

/approve

EdDev commented:
/test pull-kubevirt-build
/test pull-kubevirt-generate 
/test pull-kubevirt-unit-test
/test pull-kubevirt-code-lint
/test pull-kubevirt-e2e-k8s-1.31-sig-network

EdDev commented:
/test all

fossedihelm reviewed (approved):
/retest
